[DOC]optimization concepts-messaging retry letter topic (#14498)
apache/pulsar: site2/docs/concepts-messaging.md

consumer.acknowledge(message);
```

> If you want a more flexible redelivery strategy, a better way is to use the **redelivery backoff mechanism**. You can redeliver messages with different delays by setting the number of times the messages are retried.

Reviewer: ```suggestion
To redeliver messages with different delays, you can use the **redelivery backoff mechanism** by setting the number of retries to deliver the messages.
```

consumer.acknowledge(message);
```

### Retry letter topic

> Retry letter topic allows you to re-consuming message even some messages are not consumed successfully. The messages that are failed to be consumed are stored in a specific topic, you can custom retry delay time for the message to be retried, and subscribers will automatically subscribe to the retry letter topic and re-consume it, when the specified maximum number of retries is reached, the message will send to a dead letter topic for manual processing.

Reviewer: ```suggestion
Retry letter topic allows you to store the messages that failed to be consumed and retry consuming them later. With this method, you can custom the interval to retry the delivery of the messages, and subscribers automatically subscribe to the retry letter topic and re-consume it. When the specified maximum number of retries is reached, the messages are moved to a dead letter topic for manual processing.
```

Reviewer: Suggest we add an illustration to demonstrate the flow of the retry letter topic and describe its difference with the delayed message queue.

```
The diagram below illustrates the concept of the retry letter topic.
[Placeholder of Image]

The intention of using retry letter topic is different from using [delayed message delivery](#delayed-message-delivery), even though both are aiming to consume a message later. Retry letter topic serves failure handling through message redelivery to ensure critical data is not lost, while delayed message delivery is intended to deliver a message with a specified time of delay.
```

By default, automatic retry is disabled. You can set `enableRetry` to `true` to enable automatic retry on the consumer.

> Use the following API to consume messages from a retry letter topic.

Reviewer: ```suggestion
Use the following API to consume messages from a retry letter topic. When the value of `maxRedeliverCount` is reached, the unconsumed messages are moved to a dead letter topic.
```

```java
Consumer<byte[]> consumer = pulsarClient.newConsumer(Schema.BYTES)

>                 .topic(topic)

Reviewer: ```suggestion
                .topic("my-topic")
```

                .subscriptionName("my-subscription")
                .subscriptionType(SubscriptionType.Shared)
                .enableRetry(true)
                .deadLetterPolicy(DeadLetterPolicy.builder()
                        .maxRedeliverCount(maxRedeliveryCount)
                        .build())
                .subscribe();
```
The default retry letter topic uses this format:
```
<topicname>-<subscriptionname>-RETRY
```
Use the Java client to specify the name of the retry letter topic.
```java
Consumer<byte[]> consumer = pulsarClient.newConsumer(Schema.BYTES)

>         .topic(topic)

Reviewer: ```suggestion
        .topic("my-topic")
```

        .subscriptionName("my-subscription")
        .subscriptionType(SubscriptionType.Shared)
        .enableRetry(true)
        .deadLetterPolicy(DeadLetterPolicy.builder()
                .maxRedeliverCount(maxRedeliveryCount)

>                 .retryLetterTopic("your-topic-name")

Reviewer: ```suggestion
                .retryLetterTopic("my-retry-letter-topic-name")
```

                .build())
        .subscribe();
```

The messages in the retry letter topic contain some special properties that are automatically created by the client.

Special property | Description
:--------------------|:-----------
`REAL_TOPIC` | The real topic name.
`ORIGIN_MESSAGE_ID` | The origin message ID. It is crucial for message tracking.

> `RECONSUMETIMES`   | The retry consume times.

Reviewer: ```suggestion
`RECONSUMETIMES`   | The number of retries to consume messages.
```

> `DELAY_TIME`      | Message delay time.

Reviewer: ```suggestion
`DELAY_TIME`      | Message delay time in milliseconds.
```

**Example**
```
REAL_TOPIC = persistent://public/default/my-topic
ORIGIN_MESSAGE_ID = 1:0:-1:0
RECONSUMETIMES = 6
DELAY_TIME = 3000
```

Use the following API to store the messages in a retrial queue.

```java
consumer.reconsumeLater(msg, 3, TimeUnit.SECONDS);
```

> Use the following API to add custom properties for the `reconsumeLater` function, at the next re-consume, it can be get from message#getProperty.

Reviewer: What does "at the next re-consume, it can be get from message#getProperty" mean?
And does "it" refer to "custom properties"?

```java
Map<String, String> customProperties = new HashMap<String, String>();
customProperties.put("custom-key-1", "custom-value-1");
customProperties.put("custom-key-2", "custom-value-2");
consumer.reconsumeLater(msg, customProperties, 3, TimeUnit.SECONDS);
```
> **Note**    
> *  Currently, retry letter topic is enabled in Shared subscription types.

> > *  Compared with negative acknowledgement, reconsumeLater suitable for scenarios that require a large number of retries and custom retry delay time.

Reviewer: ```suggestion
> *  Compared with negative acknowledgment, retry letter topic is more suitable for messages that require a large number of retries with a configurable retry interval.
```

consumer.acknowledge(message);
```

### Retry letter topic

> Retry letter topic allows you to store the messages that failed to be consumed and retry consuming them later. With this method, you can custom the interval to retry the delivery of the messages, and subscribers automatically subscribe to the retry letter topic and re-consume it. When the specified maximum number of retries is reached, the unconsumed messages are moved to a [dead letter topic](#dead-letter-topic) for manual processing.

Reviewer 2: Please review the grammatical errors and update. Thanks!

Author: thx @david-streamlio , fixed

Special property | Description
:--------------------|:-----------
`REAL_TOPIC` | The real topic name.
`ORIGIN_MESSAGE_ID` | The origin message ID. It is crucial for message tracking.
`RECONSUMETIMES`   | The number of retries to consume messages.
`DELAY_TIME`      | Message retry interval in milliseconds.
**Example**
```
REAL_TOPIC = persistent://public/default/my-topic
ORIGIN_MESSAGE_ID = 1:0:-1:0
RECONSUMETIMES = 6
DELAY_TIME = 3000
```

Use the following API to store the messages in a retrial queue.

```java
consumer.reconsumeLater(msg, 3, TimeUnit.SECONDS);
```

Use the following API to add custom properties for the `reconsumeLater` function. In the next attempt to consume, custom properties can be get from message#getProperty.

```java
Map<String, String> customProperties = new HashMap<String, String>();
customProperties.put("custom-key-1", "custom-value-1");
customProperties.put("custom-key-2", "custom-value-2");
consumer.reconsumeLater(msg, customProperties, 3, TimeUnit.SECONDS);
```
> **Note**    
> *  Currently, retry letter topic is enabled in Shared subscription types.

> > *  Compared with negative acknowledgment, retry letter topic is more suitable for messages that require a large number of retries with a configurable retry interval, because the negative acknowledgment caches messages that need to be retried on the client side, and the retry letter topic is persisted to bookkeeper.

Reviewer: ```suggestion
> *  Compared with negative acknowledgment, retry letter topic is more suitable for messages that require a large number of retries with a configurable retry interval. Because messages in the retry letter topic are persisted to BookKeeper, while messages that need to be retried due to negative acknowledgment are cached on the client side.
```

```java
Consumer<byte[]> consumer = pulsarClient.newConsumer(Schema.BYTES)
        .topic(my-topic)
        .subscriptionName("my-subscription")
        .subscriptionType(SubscriptionType.Shared)
        .enableRetry(true)
        .deadLetterPolicy(DeadLetterPolicy.builder()
                .maxRedeliverCount(maxRedeliveryCount)
                .retryLetterTopic("my-retry-letter-topic-name")
                .build())
        .subscribe();
```

The messages in the retry letter topic contain some special properties that are automatically created by the client.

Special property | Description
:--------------------|:-----------
`REAL_TOPIC` | The real topic name.
`ORIGIN_MESSAGE_ID` | The origin message ID. It is crucial for message tracking.
`RECONSUMETIMES`   | The number of retries to consume messages.
`DELAY_TIME`      | Message retry interval in milliseconds.
**Example**
```
REAL_TOPIC = persistent://public/default/my-topic
ORIGIN_MESSAGE_ID = 1:0:-1:0
RECONSUMETIMES = 6
DELAY_TIME = 3000
```

Use the following API to store the messages in a retrial queue.

```java
consumer.reconsumeLater(msg, 3, TimeUnit.SECONDS);
```

> Use the following API to add custom properties for the `reconsumeLater` function, at the next re-consume, custom properties can be get from message#getProperty.

Reviewer: ```suggestion
Use the following API to add custom properties for the `reconsumeLater` function. In the next attempt to consume, custom properties can be get from message#getProperty.
```

                
```

> Dead letter topic depends on message redelivery. Messages are redelivered either due to [acknowledgement timeout](#acknowledgement-timeout) or [negative acknowledgement](#negative-acknowledgement) or [reconsumeLater](#retry-letter-topic) . If you are going to use negative acknowledgement on a message, make sure it is negatively acknowledged before the acknowledgement timeout. 

Reviewer: The actual dependency is more like "message redelivery depends on dead letter topic" or "Dead letter topic serves message redelivery"? 

```suggestion
Dead letter topic serves message redelivery, which is triggered by [acknowledgement timeout](#acknowledgement-timeout) or [negative acknowledgement](#negative-acknowledgement) or [retry letter topic](#retry-letter-topic) . If you are going to use negative acknowledgment on a message, make sure it is negatively acknowledged before the acknowledgment timeout. 

```

I think the last sentence is better to be moved to somewhere in the "negative acknowledgment" section.

> **Note**    
>  * Currently, dead letter topic is enabled in Shared and Key_Shared subscription types.

> >  * Compared with dead letter topic, retry letter topic is transparent to the user, automatically subscribed and reversed by consumers.

Reviewer: Since the subject of this sentence is "retry letter topic", does it make sense to move it to the "retry letter topic" section?

- [Negative Acknowledgment](#negative-acknowledgement)
- [Acknowledgement Timeout](#acknowledgement-timeout)

> - [ReconsumeLater](#retry-letter-topic)

Reviewer: I suggest we keep using the "Retry letter topic" term to keep consistency and avoid back-and-forth thinking.
